### PR TITLE
`azurerm_kusto_attached_database_configuration` - Add support for `database_name_override`, `database_name_prefix`, `functions_to_exclude`, and `functions_to_include` properties

### PR DESCRIPTION
**Pull request by liuwuliuyun**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR adds support for the following new properties to the `azurerm_kusto_attached_database_configuration` resource:

**Top-level properties:**
- `database_name_override` - Overrides the original database name. Relevant only when attaching to a specific database.
- `database_name_prefix` - Adds a prefix to the attached databases name. When following an entire cluster, that prefix would be added to all of the databases' original names from the leader cluster.

**Properties in the `sharing` block:**
- `functions_to_exclude` - List of functions to exclude from the follower database.
- `functions_to_include` - List of functions to include in the follower database.

These properties are already supported by the Azure API (2024-04-13) and the underlying SDK but were not exposed in the Terraform resource.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: "`resource_name_here` - description of change e.g. adding property `new_property_name_here`"

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

### New Test Cases Added:
- `TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride` - Tests the `database_name_override` property
- `TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix` - Tests the `database_name_prefix` property with `database_name = "*"`

```
=== RUN   TestAccKustoAttachedDatabaseConfiguration_basic
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_basic
=== CONT  TestAccKustoAttachedDatabaseConfiguration_basic
--- PASS: TestAccKustoAttachedDatabaseConfiguration_basic (1326.36s)
PASS

Process finished with the exit code 0

=== RUN   TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride
=== CONT  TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride
--- PASS: TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride (1223.11s)
PASS


Process finished with the exit code 0

=== RUN   TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix
=== CONT  TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix
--- PASS: TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix (1355.98s)
PASS


Process finished with the exit code 0

=== RUN   TestAccKustoAttachedDatabaseConfiguration_clusterResourceId
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_clusterResourceId
=== CONT  TestAccKustoAttachedDatabaseConfiguration_clusterResourceId
--- PASS: TestAccKustoAttachedDatabaseConfiguration_clusterResourceId (1264.74s)
PASS


Process finished with the exit code 0
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-merging.md).

* `azurerm_kusto_attached_database_configuration` - Add support for `database_name_override`, `database_name_prefix`, `functions_to_exclude`, and `functions_to_include` properties [GH-31470]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## AI Assistance Disclosure

- [x] AI Assisted - This contribution was made by, or with the assistance of, AI/LLMs

This PR was created with the assistance of GitHub Copilot for code generation, test cases, and documentation updates.

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

No changes to security controls in this pull request.

### CODE REVIEW

File: internal/services/kusto/kusto_attached_database_configuration_resource.go [hunk 1]
```diff
@@ -119,6 +131,22 @@ func resourceKustoAttachedDatabaseConfiguration() *pluginsdk.Resource {
 							},
 						},
 
+						"functions_to_exclude": {
+							Type:     pluginsdk.TypeSet,
```
Comment (ziyeqf) [inline internal/services/kusto/kusto_attached_database_configuration_resource.go:148]: just to confirm, is it possible to use `TypeList` since `TypeSet` may cause trouble in the future

**Comment by ziyeqf** [comment]
Thanks @liuwuliuyun, just one minor question, mostly LGTM

File: website/docs/r/kusto_attached_database_configuration.html.markdown
Comment (magodo) [inline website/docs/r/kusto_attached_database_configuration.html.markdown]:
```suggestion
  cluster_id = azurerm_kusto_cluster.followed_cluster.id
```

File: internal/services/kusto/kusto_attached_database_configuration_resource.go [hunk 2]
```diff
@@ -246,6 +274,8 @@ func resourceKustoAttachedDatabaseConfigurationRead(d *pluginsdk.ResourceData, m
 			}
 			d.Set("cluster_id", clusterResourceId.ID())
 			d.Set("database_name", props.DatabaseName)
+			d.Set("database_name_override", props.DatabaseNameOverride)
+			d.Set("database_name_prefix", props.DatabaseNamePrefix)
```
Comment (magodo) [inline internal/services/kusto/kusto_attached_database_configuration_resource.go]: These shall be wrapped with `pointer.From()`.

File: website/docs/r/kusto_attached_database_configuration.html.markdown [hunk 3]
```diff
@@ -89,6 +91,10 @@ The following arguments are supported:
 
 * `database_name` - (Required) The name of the database which you would like to attach, use * if you want to follow all current and future databases. Changing this forces a new resource to be created.
 
+* `database_name_override` - (Optional) Overrides the original database name. Relevant only when attaching to a specific database.
```
Comment (magodo) [inline website/docs/r/kusto_attached_database_configuration.html.markdown]: Can we reword it:


```suggestion
* `database_name_override` - (Optional) The database name to use for the attached database instead of using the original database name. Relevant only when attaching to a specific database.
```

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: internal/services/kusto/kusto_attached_database_configuration_resource.go (hunk 1 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource.go:148]: I use `TypeSet` here to maintain consistency with the other subproperties in this block. Since this property does not have a specific order, `TypeSet` is appropriate. If we decide to make changes in the future, we can update the entire block at once.

File: internal/services/kusto/kusto_attached_database_configuration_resource.go (hunk 2 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource.go]: Changed according to the comment

**Review by magodo — DISMISSED** [review]

File: internal/services/kusto/kusto_attached_database_configuration_resource.go [hunk 4]
```diff
@@ -82,6 +82,18 @@ func resourceKustoAttachedDatabaseConfiguration() *pluginsdk.Resource {
 				ValidateFunc: commonids.ValidateKustoClusterID,
 			},
 
+			"database_name_override": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource.go]: My assumption is this will have similar limitations as the `database_name` field (minus the asterisk `*` as a valid option), validation should be improved. Please confirm and update.

As an example, there definitely are characters that are not supported:

```
│ Status: "Failed"
│ Code: "BadRequest_EntityNameIsNotValid"
│ Message: "[BadRequest] 'myspecialName@!@#%^^&': 'Name' is not valid for entities of kind 'DatabaseName'. Reason='DatabaseName: Name='myspecialName@!@#%^^&' does not comply with naming rules (contains invalid characters or format mismatch)'"
```

[reference](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/schema-design-considerations.md#validation)

File: internal/services/kusto/kusto_attached_database_configuration_resource.go [hunk 5]
```diff
@@ -82,6 +82,18 @@ func resourceKustoAttachedDatabaseConfiguration() *pluginsdk.Resource {
...
+			},
+
+			"database_name_prefix": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource.go]: Similar to my comment above, this needs better validation


```
│ Status: "Failed"
│ Code: "BadRequest_EntityNameIsNotValid"
│ Message: "[BadRequest] 'myspecialPrefix@!@#%^^&acctestkd2-0217': 'Name' is not valid for entities of kind 'DatabaseName'. Reason='DatabaseName: Name='myspecialPrefix@!@#%^^&acctestkd2-0217' does not comply with naming rules (contains invalid characters or format mismatch)'"
```

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go [hunk 6]
```diff
@@ -33,6 +33,36 @@ func TestAccKustoAttachedDatabaseConfiguration_basic(t *testing.T) {
 	})
 }
 
+func TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride(t *testing.T) {
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource_test.go:36]: Please add an update test as well, none of the tests actually test the update path for any properties in this resource.

[reference](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-acceptance-testing.md#which-tests-are-required)

File: internal/services/kusto/kusto_attached_database_configuration_resource.go [hunk 7]
```diff
@@ -82,6 +82,18 @@ func resourceKustoAttachedDatabaseConfiguration() *pluginsdk.Resource {
 				ValidateFunc: commonids.ValidateKustoClusterID,
 			},
 
+			"database_name_override": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"database_name_prefix": {
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource.go:103]: Please add a `ConflictsWith` on these, you cannot set both at the same time:

```
unexpected status 400 (400 Bad Request) with error: DatabaseNameOverrideAndPrefixBadInput: Request cannot have both Database Name Override and Database Name Prefix. Please mention only one of them
```

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go [hunk 8]
```diff
@@ -129,6 +159,8 @@ resource "azurerm_kusto_attached_database_configuration" "test" {
   sharing {
     external_tables_to_exclude    = ["ExternalTable2"]
     external_tables_to_include    = ["ExternalTable1"]
+    functions_to_exclude          = ["Function2"]
+    functions_to_include          = ["Function1"]
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource_test.go:142]: When adding the update test, please ensure there is a step that has multiple strings in the set

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go [hunk 9]
```diff
@@ -141,7 +173,11 @@ resource "azurerm_kusto_attached_database_configuration" "test" {
 func (KustoAttachedDatabaseConfigurationResource) clusterResourceId(data acceptance.TestData) string {
 	return fmt.Sprintf(`
 provider "azurerm" {
-  features {}
+  features {
+    resource_group {
+      prevent_deletion_if_contains_resources = false
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource_test.go]: Why is this needed?

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go [hunk 10]
```diff
@@ -204,3 +240,129 @@ resource "azurerm_kusto_attached_database_configuration" "test" {
...
+  location            = azurerm_resource_group.rg.location
+  cluster_name        = azurerm_kusto_cluster.cluster1.name
+}
+
+resource "azurerm_kusto_database" "test" {
+  name                = "acctestkd2-%d"
+  resource_group_name = azurerm_resource_group.rg.name
+  location            = azurerm_resource_group.rg.location
+  cluster_name        = azurerm_kusto_cluster.cluster2.name
+}
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource_test.go]: Can you add a `template()` function that contains all of these dependencies so there's no need to copy-paste the same config

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go [hunk 11]
```diff
@@ -204,3 +240,129 @@ resource "azurerm_kusto_attached_database_configuration" "test" {
...
+`, data.RandomInteger, data.Locations.Primary, data.RandomString, data.RandomString, data.RandomInteger, data.RandomInteger, data.RandomInteger)
+}
+
+func (KustoAttachedDatabaseConfigurationResource) databaseNamePrefix(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {
+    resource_group {
+      prevent_deletion_if_contains_resources = false
```
Comment (sreallymatt) [inline internal/services/kusto/kusto_attached_database_configuration_resource_test.go]: Same as above, why is this needed?

File: website/docs/r/kusto_attached_database_configuration.html.markdown [hunk 12]
```diff
@@ -89,6 +91,10 @@ The following arguments are supported:
 
 * `database_name` - (Required) The name of the database which you would like to attach, use * if you want to follow all current and future databases. Changing this forces a new resource to be created.
 
+* `database_name_override` - (Optional) The database name to use for the attached database instead of using the original database name. Relevant only when attaching to a specific database.
```
Comment (sreallymatt) [inline website/docs/r/kusto_attached_database_configuration.html.markdown:94]: `Relevant only when attaching to a specific database` - based on this, I assume `database_name_override` is invalid if `database_name` is `*`. We should validate for that.

**Review by sreallymatt — CHANGES_REQUESTED** [review]

File: internal/services/kusto/kusto_attached_database_configuration_resource.go (hunk 4 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource.go]: Good point, will add validation for database name and prefix

File: internal/services/kusto/kusto_attached_database_configuration_resource.go (hunk 5 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource.go]: Same as above

File: internal/services/kusto/kusto_attached_database_configuration_resource.go (hunk 7 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource.go:103]: Will add conflictsWith here

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go (hunk 9 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource_test.go]: This process is a bit complex. Creating and deleting a Kusto cluster and its database can take over two hours. Even after the resource is deleted, it may still appear in the resource group list due to a cache consistency issue in ARM, which can cause the process to timeout repeatedly. If you want the acctest to run until completion, you’ll need to add this flag.

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go (hunk 10 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource_test.go]: Will do

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go (hunk 11 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource_test.go]: Same above

File: website/docs/r/kusto_attached_database_configuration.html.markdown (hunk 12 above)
Comment (liuwuliuyun) [reply website/docs/r/kusto_attached_database_configuration.html.markdown:94]: Yeah, I will validate this

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go (hunk 8 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource_test.go:142]: Okay, will try

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go (hunk 6 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource_test.go:36]: Will try.

File: internal/services/kusto/kusto_attached_database_configuration_resource_test.go (hunk 9 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_attached_database_configuration_resource_test.go]: I will try to remove this and run tests locally to see if it works.

**Comment by liuwuliuyun** [comment]
Latest testing results:

```
PS C:\Users\yunliu1\Documents\terraform-provider-azurerm> go test -v -timeout 120m -run "TestAccKustoAttachedDatabaseConfiguration_basic" ./internal/services/kusto/
=== RUN   TestAccKustoAttachedDatabaseConfiguration_basic
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_basic
=== CONT  TestAccKustoAttachedDatabaseConfiguration_basic
--- PASS: TestAccKustoAttachedDatabaseConfiguration_basic (1484.27s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/kusto 1502.622s
PS C:\Users\yunliu1\Documents\terraform-provider-azurerm> go test -v -timeout 120m -run "TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride" ./internal/services/kusto/
=== RUN   TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride
=== CONT  TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride
--- PASS: TestAccKustoAttachedDatabaseConfiguration_databaseNameOverride (1349.41s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/kusto 1369.040s
PS C:\Users\yunliu1\Documents\terraform-provider-azurerm> go test -v -timeout 120m -run "TestAccKustoAttachedDatabaseConfiguration_update" ./internal/services/kusto/              
=== RUN   TestAccKustoAttachedDatabaseConfiguration_update
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_update
=== CONT  TestAccKustoAttachedDatabaseConfiguration_update
--- PASS: TestAccKustoAttachedDatabaseConfiguration_update (1559.69s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/kusto 1567.957s
PS C:\Users\yunliu1\Documents\terraform-provider-azurerm> go test -v -timeout 120m -run "TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix" ./internal/services/kusto/
=== RUN   TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix
=== CONT  TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix
--- PASS: TestAccKustoAttachedDatabaseConfiguration_databaseNamePrefix (1307.91s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/kusto 1316.266s
PS C:\Users\yunliu1\Documents\terraform-provider-azurerm> go test -v -timeout 120m -run "TestAccKustoAttachedDatabaseConfiguration_clusterResourceId" ./internal/services/kusto/
=== RUN   TestAccKustoAttachedDatabaseConfiguration_clusterResourceId
=== PAUSE TestAccKustoAttachedDatabaseConfiguration_clusterResourceId
=== CONT  TestAccKustoAttachedDatabaseConfiguration_clusterResourceId
--- PASS: TestAccKustoAttachedDatabaseConfiguration_clusterResourceId (1284.11s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/kusto 1291.225s
```

**Comment by liuwuliuyun** [comment]
Hi @sreallymatt , I've addressed all the review feedback. Here's a summary of the changes:

1. **Validation for `database_name_override` and `database_name_prefix`** — Replaced `validation.StringIsNotEmpty` with `validate.DatabaseName` on both fields, which enforces the Kusto database naming rules (alphanumeric, whitespace, dashes, underscores, dots only).

2. **`ConflictsWith` on `database_name_override` / `database_name_prefix`** — Added `ConflictsWith` on both fields so they cannot be set simultaneously, and added a corresponding note in the documentation.

3. **`database_name_override` invalid when `database_name` is `*`** — Added a `CustomizeDiff` that returns an error if `database_name_override` is set while `database_name` is `*`.

4. **Update test** — Added `TestAccKustoAttachedDatabaseConfiguration_update` that goes from `basic` → `complete`, where the `complete` config includes multiple strings in each set field (e.g., `["Function1", "Function2"]`).

5. **Shared `template()` function** — Extracted all common dependencies (resource group, clusters, databases) into a `template()` function to eliminate copy-paste across test configs.

6. **Removed `prevent_deletion_if_contains_resources = false`** — Removed the unnecessary provider feature flag from the test configs.

All tests pass locally. Could you please take another look when you get a chance? Thanks!

**Review by sreallymatt — APPROVED** [review]
Thanks @liuwuliuyun - LGTM ✅
